File: backend/app/services/context.py
"""
Everything the sprint holds for a phase, assembled in the order the agents
should weigh it.

This is what makes each phase a continuation rather than a fresh start: the
engagement and its destination, who was in the room, what the client could not
supply, the notes typed by hand, the data room, earlier deliverables, the two
previous outputs, and every question already put to the client — with its
answer where one came back.
"""

from __future__ import annotations

import json
from typing import Any

from app.domain import playbook


def _scope_rule(scope: str) -> str:
    if "process" in (scope or "").lower():
        return (
            "Sprint scope: SINGLE PROCESS-LEVEL SPRINT. The unit of analysis is ONE named process "
            "end to end. Diagnose it step by step - every step, handoff, system, wait, exception "
            "and rework loop in the order work actually moves - and quantify at step level. Do not "
            "spread the analysis across departments or rank functions against each other; where a "
            "neighbouring function appears, treat it only as an input or output of this process."
        )
    return (
        "Sprint scope: DEPARTMENT-LEVEL SPRINT. The unit of analysis is a whole function or "
        "department. Diagnose ACROSS the processes it owns - compare them, size the value in each, "
        "and rank them - so the sprint can select which process to take into forensics. Do not "
        "collapse the work into a single process."
    )


def build(
    *,
    engagement: dict[str, Any],
    phase: int,
    state: dict[str, Any],
    packs: dict[int, Any],
    questions: list[dict[str, Any]],
    include_room: bool = True,
) -> str:
    lines: list[str] = [
        f"Client: {engagement['name']}",
        f"Sector: {engagement['sector']}",
        f"Website: {engagement.get('url') or 'not supplied'}",
        _scope_rule(engagement.get("scope", "")),
        f"Engagement notes: {engagement.get('notes') or 'none'}",
    ]

    if engagement.get("brief"):
        lines.append(f"What this sprint is for, in the consultant's words: {engagement['brief']}")

    metrics = engagement.get("success_metrics") or []
    if metrics:
        lines.append(
            "The destination this sprint is measured against, captured before any phase ran:\n"
            + "\n".join(
                f"- {m['metric']}: {m.get('baseline') or '?'} → {m.get('target') or '?'}"
                f"{' ' + m['horizon'] if m.get('horizon') else ''}"
                f"{'  [PRIMARY]' if m.get('is_primary') else ''}"
                for m in metrics
            )
        )

    links = state.get("links") or []
    if links:
        lines.append("Pasted source links: " + "; ".join(x["url"] for x in links))

    # inputs the client could not supply
    na: list[str] = []
    for row in state.get("inputs", []):
        if row["state"] != "na" or row["phase"] > phase:
            continue
        ph = playbook.phase(row["phase"])
        if row["input_index"] < len(ph["inputs"]):
            na.append(f"[P{ph['num']}] {ph['inputs'][row['input_index']]}")
    if na:
        lines.append(
            "Inputs the client could not supply - benchmark these against the sector and mark the "
            "figure as derived:\n" + "\n".join(f"- {x}" for x in na)
        )

    # who was in the room
    attended = {(r["phase"], r["participant_index"]) for r in state.get("attendance", [])}
    rooms: list[str] = []
    for i in range(phase + 1):
        ph = playbook.phase(i)
        expected = [(j, w) for j, w in enumerate(ph["participants"]) if not w.startswith("No broad")]
        if not expected:
            continue
        present = [w for j, w in expected if (i, j) in attended]
        missing = [w for j, w in expected if (i, j) not in attended]
        rooms.append(
            f"[Phase {ph['num']}] present: {'; '.join(present) if present else 'not recorded'}"
            + (f"\n  not in the room: {'; '.join(missing)}" if missing else "")
        )
    if rooms:
        lines.append(
            "Who was in the room for each phase. Where someone the playbook asks for was absent, "
            "say plainly which conclusions rest on an unverified view rather than presenting them "
            "as settled:\n" + "\n".join(rooms)
        )

    # notes typed by the consultant
    notes = [r for r in state.get("notes", []) if r["phase"] <= phase and r["body"]]
    if notes:
        lines.append(
            "Manually entered information:\n"
            + "\n\n".join(
                f"[Phase {playbook.phase(r['phase'])['num']} - entered by the consultant, treat as "
                f"first-hand client evidence]\n{r['body']}"
                for r in sorted(notes, key=lambda r: r["phase"])
            )
        )

    files = [f for f in state.get("files", []) if f["phase"] <= phase] if include_room else []
    if files:
        lines.append(
            "Data room: "
            + "; ".join(
                f"{f['name']}{' [text read]' if f.get('extracted_text') else ' [contents not parsed]'}"
                for f in files
            )
        )
        parsed = [f for f in files if f.get("extracted_text")]
        if parsed:
            lines.append(
                "Verbatim extracts from the uploaded files and meeting transcripts - treat these as "
                "first-hand client evidence, and note any question already put to the client inside "
                "them:\n" + "\n\n".join(f"[{f['name']}]\n{f['extracted_text'][:6000]}" for f in parsed[:4])
            )

    # earlier deliverable drafts
    priors = [
        d for d in state.get("deliverables", []) if d["phase"] < phase and (d.get("draft") or "").strip()
    ]
    if priors:
        lines.append(
            "Earlier phase deliverables (extracts):\n"
            + "\n".join(
                f"- [Phase {playbook.phase(d['phase'])['num']}] "
                f"{playbook.phase(d['phase'])['docs'][d['doc_number'] - 1]['name']}: {d['draft'][:900]}"
                for d in priors[:6]
            )
        )

    """
    Every earlier pack, not just the last two.

    Each phase is built on the ones before it, so a phase that cannot see Phase
    0's value tree while writing a Phase 5 business case will quietly reinvent
    it. The two nearest phases are carried at full width because this phase
    argues directly with them; the earlier ones are carried narrower so the
    whole chain fits without crowding out the live sources.
    """
    carried = []
    for i in range(phase):
        if i not in packs:
            continue
        width = 4000 if i >= phase - 2 else 1600
        carried.append(
            f"[Phase {playbook.phase(i)['num']} - {playbook.phase(i)['title']} output]\n"
            f"{json.dumps(packs[i])[:width]}"
        )
    if carried:
        lines.append(
            "EVERY OUTPUT PACK BUILT BEFORE THIS PHASE, oldest first. This phase is a continuation "
            "of them, not a fresh start: carry their figures forward, name the priorities they "
            "selected, reuse their definitions, and never contradict one without saying which new "
            "evidence moved it:\n" + "\n\n".join(carried)
        )

    # questions, with answers where they came back
    asked = [q for q in questions if q["kind"] == "open" and q["phase"] < phase]
    if asked:
        answered = [q for q in asked if q.get("answered_at")]
        still_open = [q for q in asked if not q.get("answered_at")]
        if answered:
            lines.append(
                "Answers the client has given to earlier questions. These are first-hand evidence - "
                "prefer them over any benchmark:\n"
                + "\n".join(
                    f"- [P{playbook.phase(q['phase'])['num']}] {q['body']}\n  ANSWER: {q['answer']}"
                    for q in answered
                )
            )
        if still_open:
            lines.append(
                "Questions put to the client at the end of the earlier phases and still unanswered. "
                "Do not assume they were answered:\n"
                + "\n".join(f"- [P{playbook.phase(q['phase'])['num']}] {q['body']}" for q in still_open)
            )

    return "\n".join(lines)


def digest(text: str) -> str:
    """A stable fingerprint of the context, so a regenerate can say nothing changed."""
    import hashlib

    return hashlib.sha256(text.encode()).hexdigest()[:32]

File: Frontend/src/lib/ai/context.ts
import { PHASES } from '@/lib/playbook/phases';
import { PACK_KEYS } from '@/lib/playbook/constants';
import type { Engagement } from '@/lib/domain/types';

/** How the sprint's scope changes the unit of analysis for every phase. */
function scopeRule(scope: string): string {
  return /process/i.test(scope)
    ? 'Sprint scope: SINGLE PROCESS-LEVEL SPRINT. The unit of analysis is ONE named process end to end (for example work order generation or PO creation). Diagnose it step by step - every step, handoff, system, wait, exception and rework loop in the order work actually moves - and quantify at step level. Do not spread the analysis across departments or rank functions against each other; where a neighbouring function appears, treat it only as an input or output of this process.'
    : 'Sprint scope: DEPARTMENT-LEVEL SPRINT. The unit of analysis is a whole function or department (for example materials management, procurement or finance). Diagnose ACROSS the processes it owns - compare them, size the value in each, and rank them - so the sprint can select which process to take into forensics. Do not collapse the work into a single process.';
}

/**
 * Everything the sprint holds for a phase, in the order the model should weigh
 * it: the engagement itself, what the client could not supply, notes typed by
 * the consultant, the data room, research, earlier deliverables, the earlier
 * output packs and the questions already put to the client.
 *
 * `includeRoom: false` is used by the research desk when the consultant wants a
 * brief written without the uploaded files.
 */
export function buildContext(p: Engagement, pi: number, includeRoom = true): string {
  const scope = p.scope || 'Department-level sprint';
  const lines: string[] = [
    'Client: ' + p.name,
    'Sector: ' + p.sector,
    'Website: ' + (p.url || 'not supplied'),
    scopeRule(scope),
    'Engagement notes: ' + (p.notes || 'none'),
  ];

  if ((p.links || []).length) lines.push('Pasted source links: ' + p.links.map((l) => l.url).join('; '));

  const naList: string[] = [];
  PHASES.forEach((ph, i) => {
    if (i > pi) return;
    ph.inputs.forEach((label, j) => {
      if (p.inputs[i + ':' + j] === 'na') naList.push('[P' + ph.num + '] ' + label);
    });
  });
  if (naList.length) {
    lines.push(
      'Inputs the client could not supply - benchmark these against the sector and mark the figure as derived:\n' +
        naList.map((x) => '- ' + x).join('\n'),
    );
  }

  /*
   * Who was actually in the room. The playbook names the participants each
   * phase needs; a missing one is a hole in the evidence, so the pack is told
   * whose view it does not have rather than writing as if it had everyone.
   */
  const rooms: string[] = [];
  PHASES.forEach((ph, i) => {
    if (i > pi) return;
    const expected = ph.participants.filter((who) => !/^No broad/i.test(who));
    if (!expected.length) return;
    const present = expected.filter((who) => p.attended?.[i + ':' + ph.participants.indexOf(who)] === true);
    const missing = expected.filter((who) => !present.includes(who));
    rooms.push(
      '[Phase ' + ph.num + '] present: ' + (present.length ? present.join('; ') : 'not recorded') +
        (missing.length ? '\n  not in the room: ' + missing.join('; ') : ''),
    );
  });
  if (rooms.length) {
    lines.push(
      'Who was in the room for each phase. Where someone the playbook asks for was absent, say plainly which conclusions rest on an unverified view rather than presenting them as settled:\n' +
        rooms.join('\n'),
    );
  }

  const manual: string[] = [];
  PHASES.forEach((ph, i) => {
    if (i > pi) return;
    const t = (p.manual || {})[i];
    if (t) manual.push('[Phase ' + ph.num + ' - entered manually by the consultant, treat as first-hand client evidence]\n' + t);
  });
  if (manual.length) lines.push('Manually entered information:\n' + manual.join('\n\n'));

  const files = includeRoom === false ? [] : p.files.filter((f) => f.phase <= pi);
  if (files.length) {
    lines.push('Data room: ' + files.map((f) => f.name + (f.txt ? ' [text read]' : ' [contents not parsed]')).join('; '));
  }
  const parsed = files.filter((f) => f.txt);
  if (parsed.length) {
    lines.push(
      'Verbatim extracts from the uploaded text files and meeting transcripts - treat these as first-hand client evidence, and note any question already put to the client inside them:\n' +
        parsed
          .slice(0, 4)
          .map((f) => '[' + f.name + ']\n' + String(f.txt).slice(0, 6000))
          .join('\n\n'),
    );
  }

  const used = (p.research || []).filter((r) => r.use !== false);
  if (used.length) {
    lines.push(
      'Research library extracts:\n' +
        used
          .slice(0, 3)
          .map((r) => '- ' + r.q + '\n' + String(r.md).slice(0, 2200))
          .join('\n\n'),
    );
  }

  const priors: string[] = [];
  PHASES.forEach((ph, i) => {
    if (i >= pi) return;
    ph.docs.forEach((d) => {
      const rec = p.docs[i + '.' + d.n];
      if (rec && rec.draft) priors.push('- [Phase ' + ph.num + '] ' + d.name + ': ' + String(rec.draft).slice(0, 900));
    });
  });
  if (priors.length) lines.push('Earlier phase deliverables (extracts):\n' + priors.slice(0, 6).join('\n'));

  const packs: string[] = [];
  for (let i = Math.max(0, pi - 2); i < pi; i++) {
    const w = p[PACK_KEYS[i]];
    if (w) packs.push('[Phase ' + PHASES[i].num + ' - ' + PHASES[i].title + ' output pack]\n' + JSON.stringify(w).slice(0, 2500));
  }
  if (packs.length) {
    lines.push(
      'Output packs already built in the phases before this one. This phase is a continuation of them: carry their figures forward, name the priorities they selected, and never contradict them without saying why:\n' +
        packs.join('\n\n'),
    );
  }

  const openQs: string[] = [];
  for (let i = 0; i < pi; i++) {
    const ent = (p.cq || {})[i];
    if (ent && (ent.items || []).length) {
      openQs.push('[Phase ' + PHASES[i].num + ']\n' + ent.items.map((it) => '- ' + it.q).join('\n'));
    }
  }
  if (openQs.length) {
    lines.push(
      'Questions put to the client at the end of the earlier phases. Anything the material does not answer is still open - do not assume it was answered:\n' +
        openQs.join('\n\n'),
    );
  }

  return lines.join('\n');
}

/** Text-bearing uploads visible to a phase — transcripts, notes, exports. */
export function textSources(p: Engagement, pi: number) {
  return (p.files || []).filter((f) => f.phase <= pi && f.txt);
}
